core/src/test/scala/com/mesosphere/usi/core/scalaapi/SchedulerAdapterTest.scala

```diff
@@ -0,0 +1,315 @@
+    val adapter = new SchedulerAdapter(scheduler)
+  }
+
+  "SchedulerAdapter" when {
+    "materialized as Source and Sink" should {
+      "Propagate a snapshot to the underlying scheduler" in new SchedulerFixture {
+        adapter.asSourceAndSink(SpecsSnapshot(List(podSpec), Nil))
+        specInputQueue.pull().futureValue.get._1.podSpecs.head shouldEqual podSpec
+      }
+
+      "If snapshot is not provided push an empty snapshot" in new SchedulerFixture {
+        adapter.asSourceAndSink()
+        specInputQueue.pull().futureValue.get._1 shouldEqual SpecsSnapshot.empty
+      }
+
+      "Complete the snapshot promise once it's available" in new SchedulerFixture {
+        val (snapshotF, _, _) = adapter.asSourceAndSink()
+        specInputQueue.pull().futureValue
+
+        val stateSnapshot = StateSnapshot(List(PodStatus(podSpec.id, Map.empty)), Nil, Nil, Nil)
+
+        stateOutputQueue.offer(stateSnapshot -> Source.maybe)
+
+        snapshotF.futureValue shouldEqual stateSnapshot
+      }
+
+      "Push the SpecUpdatedEvents downstream" in new SchedulerFixture {
+        val (_, _, sink) = adapter.asSourceAndSink()
+
+        Source.repeat(specEvent).runWith(sink)
+
+        specInputQueue.pull().futureValue.get._2.runWith(Sink.head).futureValue shouldEqual specEvent
+      }
+
+      "Push the events from scheduler to the provided Source" in new SchedulerFixture {
+        val (_, source, _) = adapter.asSourceAndSink()
+
+        stateOutputQueue.offer(StateSnapshot.empty -> Source.repeat(stateEvent))
+
+        source.runWith(Sink.head).futureValue shouldEqual stateEvent
+      }
+
+      "Complete both sink and source when the event stream coming to scheduler is cancelled by the scheduler" in new SchedulerFixture {
+        val (_, source, sink) = adapter.asSourceAndSink()
+        val (_, innerSource) = specInputQueue.pull().futureValue.get
+
+        val stateStreamCompletionPromise = Promise[String]()
+        val specStreamCompletionPromise = Promise[String]()
+
+        Source
+          .repeat(specEvent)
+          .watchTermination()((_, terminationSignal) =>
+            terminationSignal.map(_ => specStreamCompletionPromise.trySuccess("completed")))
+          .runWith(sink)
+        source
+          .watchTermination()((_, terminationSignal) =>
+            terminationSignal.map(_ => stateStreamCompletionPromise.trySuccess("completed")))
+          .runWith(Sink.ignore)
+
+        innerSource.runWith(Sink.cancelled) // cancel incoming updates on scheduler size
+
+        specStreamCompletionPromise.future.futureValue shouldEqual "completed"
+        stateStreamCompletionPromise.future.futureValue shouldEqual "completed"
+      }
+
+      "Complete both sink and source when the event stream coming to scheduler is cancelled by the client" in new SchedulerFixture {
+        val (_, source, sink) = adapter.asSourceAndSink()
+        val (_, innerSource) = specInputQueue.pull().futureValue.get
```

Comment:
```suggestion
        val (_, innerSource) = specInputQueue.pull().futureValue.value
```

Reply:
```
 val (_, Some(innerSource)) = specInputQueue.pull().futureValue
```

Reply:
Also, using _ in deconstruction assignment actually creates a variable with that name. Instead, use Inside